Service/Microsoft.Health.DeIdentification.Local/InMemoryQueueClient.cs

```diff
@@ -1,37 +1,106 @@
+            jobInfoStore.Result = jobInfo.Result;
+
+            if (requestCancellationOnFailure && jobInfo.Status == JobStatus.Failed)
+            {
+                await CancelJobByIdAsync(jobInfo.QueueType, jobInfo.Id, cancellationToken);
+            }
         }
 
         public Task<JobInfo> DequeueAsync(byte queueType, string worker, int heartbeatTimeoutSec, CancellationToken cancellationToken, long? jobId = null)
         {
-            throw new NotImplementedException();
+            JobInfo job = jobInfos.FirstOrDefault(t => t.Status == JobStatus.Created || (t.Status == JobStatus.Running && (DateTime.Now - t.HeartbeatDateTime) > TimeSpan.FromSeconds(heartbeatTimeoutSec)));
+            if (job != null)
+            {
+                job.Status = JobStatus.Running;
+                job.HeartbeatDateTime = DateTime.Now;
+            }
+
+            return Task.FromResult(job);
         }
 
-        public Task<IReadOnlyCollection<JobInfo>> DequeueJobsAsync(byte queueType, int numberOfJobsToDequeue, string worker, int heartbeatTimeoutSec, CancellationToken cancellationToken)
+        public async Task<IReadOnlyCollection<JobInfo>> DequeueJobsAsync(byte queueType, int numberOfJobsToDequeue, string worker, int heartbeatTimeoutSec, CancellationToken cancellationToken)
         {
-            throw new NotImplementedException();
+            var jobs = new List<JobInfo>();
+            while(numberOfJobsToDequeue > 0)
+            {
+                JobInfo job = jobInfos.FirstOrDefault(t => t.Status == JobStatus.Created || (t.Status == JobStatus.Running && (DateTime.Now - t.HeartbeatDateTime) > TimeSpan.FromSeconds(heartbeatTimeoutSec)));
+                if (job != null)
+                {
+                    job.Status = JobStatus.Running;
+                    job.HeartbeatDateTime = DateTime.Now;
+                }
+                jobs.Add(job);
+            }
+            return jobs;
         }
 
         public Task<IReadOnlyList<JobInfo>> EnqueueAsync(byte queueType, string[] definitions, long? groupId, bool forceOneActiveJobGroup, bool isCompleted, CancellationToken cancellationToken)
         {
-            throw new NotImplementedException();
+            var result = new List<JobInfo>();
+
+            long gId = groupId ?? largestId++;
+            foreach (string definition in definitions)
+            {
+                if (jobInfos.Any(t => t.Definition.Equals(definition)))
+                {
+                    result.Add(jobInfos.First(t => t.Definition.Equals(definition)));
+                    continue;
+                }
+
+                result.Add(new JobInfo()
+                {
+                    Definition = definition,
+                    Id = largestId,
+                    GroupId = gId,
+                    Status = JobStatus.Created,
+                    HeartbeatDateTime = DateTime.Now,
+                });
+                largestId++;
+            }
+
+            jobInfos.AddRange(result);
```

Comment:
Do we need ensure thread safety for enqueue/dequeue operations? Or we can simply declare this client should be used by single thread only.

Reply:
I will add TODO for it, we will check it later